src/operator/trace.rs

```diff
@@ -0,0 +1,427 @@
+    }
+
+    // TODO: this method should replace `Stream::integrate()`.
+    pub fn integrate_trace<T>(&self) -> Stream<Circuit<P>, T>
+    where
+        B: Batch,
+        T: NumEntries + DeepSizeOf + Trace<Batch = Rc<B>> + Clone + 'static,
+        T::Time: Timestamp,
+    {
+        self.circuit()
+            .cache_get_or_insert_with(IntegrateTraceId::new(self.local_node_id()), || {
+                self.circuit().region("integrate_trace", || {
+                    let (ExportStream { local, export }, z1feedback) =
+                        self.circuit().add_feedback_with_export(Z1Trace::new(true));
+                    let trace = self.circuit().add_binary_operator_with_preference(
+                        <UntimedTraceAppend<T, B>>::new(),
+                        &local,
+                        self,
+                        OwnershipPreference::STRONGLY_PREFER_OWNED,
+                        OwnershipPreference::PREFER_OWNED,
+                    );
+                    z1feedback.connect_with_preference(
+                        &trace,
+                        OwnershipPreference::STRONGLY_PREFER_OWNED,
+                    );
+                    self.circuit()
+                        .cache_insert(DelayedIntegrateTraceId::new(trace.local_node_id()), local);
+                    self.circuit()
+                        .cache_insert(ExportId::new(trace.local_node_id()), export);
+                    trace
+                })
+            })
+            .clone()
+    }
+}
+
+impl<P, T> Stream<Circuit<P>, T>
+where
+    P: Clone + 'static,
+    T: TraceReader + 'static,
+{
+    pub fn delay_trace(&self) -> Stream<Circuit<P>, T> {
+        self.circuit()
+            .cache_get_or_insert_with(
+                DelayedTraceId::new(self.local_node_id()),
+                || unimplemented!(),
+            )
+            .clone()
+    }
+}
+
+pub struct UntimedTraceAppend<T, B>
+where
+    T: TraceReader,
+{
+    _phantom: PhantomData<(T, B)>,
+}
+
+impl<T, B> UntimedTraceAppend<T, B>
+where
+    T: TraceReader,
+{
+    pub fn new() -> Self {
+        Self {
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T, B> Default for UntimedTraceAppend<T, B>
+where
+    T: TraceReader,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T, B> Operator for UntimedTraceAppend<T, B>
+where
+    T: TraceReader + 'static,
+    B: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("UntimedTraceAppend")
+    }
+    fn clock_start(&mut self, _scope: Scope) {}
+    fn clock_end(&mut self, _scope: Scope) {}
+    fn fixedpoint(&self) -> bool {
+        true
+    }
+}
+
+impl<T, B> BinaryOperator<T, B, T> for UntimedTraceAppend<T, B>
+where
+    B: Batch + Clone + 'static,
+    T: Trace<Batch = Rc<B>> + 'static,
+{
+    fn eval(&mut self, _trace: &T, _batch: &B) -> T {
+        // Refuse to accept trace by reference.  This should not happen in a correctly
+        // constructed circuit.
```

Comment:
The use of `STRONGLY_PREFER_OWNED` to connect the input of this operator guarantees that we will either get an owned value or circuit construction will fail.  The latter won't happen in practice, because this operator is only instantiated by methods in this module.